Call lift fresh in valueToTensor

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81927

`valueToTensor` is invoked in many python apis such as `x[0] = 0.5`, which lifts the 0.5 to a tensor. the problem is described similarly in https://github.com/pytorch/pytorch/pull/81609 (/s/scalar_to_tensor/valueToTensor)
> scalar_to_tensor is not dispatched and thus there is no interposition point for modes to ensure that the resulting tensor is appropriately wrapped. lift_fresh introduces this interposition point. This prevents FakeTensorMode from erroring

eellison committed:
Call lift fresh in valueToTensor

[ghstack-poisoned]

eellison commented:
cc @bdhirsh 

> RuntimeError: !at::functionalization::impl::isFunctionalTensor(self) INTERNAL ASSERT FAILED at "/var/lib/jenkins/workspace/aten/src/ATen/FunctionalizeFallbackKernel.cpp":186, please report a bug to PyTorch.

ezyang commented:
The assert failure suggests that this is not an appropriate use of `lift_fresh`. `lift_fresh` is supposed to be used in situations where you are guaranteed to get a plain Tensor. When we look at the implementation, this is true for cpu device but not for non cpu device, which the tests you're triggering (LTC device) apply

```
static inline Tensor scalarToTensor(
    const Scalar& v,
    const TensorOptions& options,
    const at::Device& self_device) {
  if (self_device == at::kCPU) {
    return at::detail::scalar_tensor_static(
        v, options.dtype_opt()->toScalarType(), self_device);
  } else {
    return impl::scalarToTensorNonNativeDeviceType(v, options);
  }
}
```

The easiest thing is probably to push `lift_fresh` into the CPU branch. Although I suppose this is a pessimization for CPU backend.

For reference, the other call dispatches

```
static inline Tensor scalarToTensorNonNativeDeviceType(
    const Scalar& v,
    const TensorOptions& options) {
  return at::scalar_tensor(v, options);
}
```

ezyang reviewed (approved):
approving with the suggested cr change

bdhirsh commented on torch/csrc/autograd/python_variable_indexing.cpp:
```diff
@@ -106,16 +106,16 @@ inline Variable valueToTensor(
   if (PyComplex_Check(value)) {
-    return at::indexing::scalarToTensor(
-        Scalar(THPUtils_unpackComplexDouble(value)), options, device);
+    return at::lift_fresh(at::indexing::scalarToTensor(
```
random side-car comment: `scalarToTensor` has a slow path that eventually calls `at::scalar_tensor`, and that op doesn't use `lift()` - I wonder if that will bite us anywhere else (basically, if there are any other user-facing API's that can cause us to hit `scalar_tensor` without going through `scalarToTensor`)

eellison committed:
Update on "Call lift fresh in valueToTensor"


`valueToTensor` is invoked in many python apis such as `x[0] = 0.5`, which lifts the 0.5 to a tensor. the problem is described similarly in https://github.com/pytorch/pytorch/pull/81609 (/s/scalar_to_tensor/valueToTensor)
> scalar_to_tensor is not dispatched and thus there is no interposition point for modes to ensure that the resulting tensor is appropriately wrapped. lift_fresh introduces this interposition point. This prevents FakeTensorMode from erroring


[ghstack-poisoned]

eellison commented:
@pytorchbot merge
